metricbeat/module/system/socket/socket.go

```diff
@@ -258,54 +263,107 @@ func newConnection(diag *linux.InetDiagMsg) *connection {
-	evt := common.MapStr{
-		"family": c.Family.String(),
-		"local": common.MapStr{
-			"ip":   c.LocalIP.String(),
-			"port": c.LocalPort,
+// Map helpers for conversion to event
+var (
+	ianaNumbersMap = map[string]string{
+		"ipv4": "4",
+		"ipv6": "41",
+	}
+
+	localHostInfoGroup = map[string]string{
+		sock.IncomingName:  "destination",
+		sock.OutgoingName:  "source",
+		sock.ListeningName: "server",
+	}
+
+	remoteHostInfoGroup = map[string]string{
+		sock.IncomingName: "source",
+		sock.OutgoingName: "destination",
+	}
+)
+
+func (c *connection) ToMapStr() (fields common.MapStr, metricSetFields common.MapStr) {
+	localGroup := "server"
+	if g, ok := localHostInfoGroup[c.Family.String()]; ok {
+		localGroup = g
+	}
+
+	fields = common.MapStr{
+		"network": common.MapStr{
+			"type":        c.Family.String(),
+			"iana_number": ianaNumbersMap[c.Family.String()],
+			"direction":   c.Direction.String(),
 		},
 		"user": common.MapStr{
 			"id": c.UID,
 		},
-		"direction": c.Direction.String(),
+		// Aliases for this are not going to be possible, keeping
+		// duplicated fields by now for backwards comatibility
+		localGroup: common.MapStr{
+			"ip":   c.LocalIP.String(),
+			"port": c.LocalPort,
+		},
+	}
+
+	metricSetFields = common.MapStr{
+		"local": common.MapStr{
+			"ip":   c.LocalIP.String(),
+			"port": c.LocalPort,
+		},
 	}
 
 	if c.Username != "" {
-		evt.Put("user.name", c.Username)
+		fields.Put("user.full_name", c.Username)
 	}
 
 	if c.ProcessError != nil {
-		evt.Put("process.error", c.ProcessError.Error())
+		// XXX, would this be the proper place for a process error code?
+		fields.Put("error.code", c.ProcessError.Error())
 	} else {
 		process := common.MapStr{"pid": c.PID}
-		evt["process"] = process
 
 		if c.PID > 0 {
-			addOptionalString(process, "exe", c.Exe)
-			addOptionalString(process, "command", c.Command)
-			addOptionalString(process, "cmdline", c.CmdLine)
+			addOptionalString(process, "executable", c.Exe)
+			addOptionalString(process, "name", c.Command)
+
+			if len(c.Args) >= 0 {
+				process["args"] = c.Args
+				metricSetFields["process"] = common.MapStr{
+					"cmdline": c.CmdLine,
+				}
+			}
 		} else if c.PID == 0 {
 			process["command"] = "kernel"
 		}
+
+		fields["process"] = process
 	}
 
 	if c.RemotePort != 0 {
+		// Aliases for this are not going to be possible, keeping
+		// duplicated fields by now for backwards comatibility
 		remote := common.MapStr{
 			"ip":   c.RemoteIP.String(),
 			"port": c.RemotePort,
 		}
-		evt["remote"] = remote
-
 		if c.DestHostError != nil {
 			remote["host_error"] = c.DestHostError.Error()
 		} else {
 			addOptionalString(remote, "host", c.DestHost)
 			addOptionalString(remote, "etld_plus_one", c.DestHostETLDPlusOne)
 		}
+		metricSetFields["remote"] = remote
+
+		remoteGroup, ok := remoteHostInfoGroup[c.Family.String()]
+		if ok {
+			fields[remoteGroup] = common.MapStr{
+				"ip":   c.RemoteIP.String(),
+				"port": c.RemotePort,
+			}
+		}
```

Comment:
This made me think to outline the exact mapping I would like to see, in terms of address/ip for the 3 possible states we can observe.

Also made me realize that ECS currently doesn't support sockets that well (noted for later) :-)

**listening**

- local.ip => server.ip
- local.port => server.port

**incoming**

- local.ip => server.ip
- local.port => server.port
- remote.ip => client.ip
- remote.port => client.port

**outgoing**

- local.ip => client.ip
- local.port => client.port
- remote.ip => server.ip
- remote.port => server.port

Reply:
For `listening` I think we can be quite sure that this is a server, but for incoming/outgoing connections I am not sure we can assume this is a client-server protocol, source/destination doesn't assumes protocols.

(Actually local/remote is the terminology used on Linux for this.)

Reply:
Yes, right now ECS defines the client/server and source/destination pairs. I think there's some contexts where local/remote would also be clearer.

If you're not comfortable using client/server in some of these situations, can you use source/destination? Note that ECS assumes that event sources fill in source/destination all of the time, and client/server only when it makes sense.

So in all the places where you have client/server, you should also duplicate these field sets to source/destination. src/dest are considered the "baseline" that you can always rely on being there.

Reply:
Yes, using source/destination now, and server only for listening sockets.

I still keep system.socket.local/remote becase we'd need N-N aliases for migration.